File: game/app/api/engine/MovableRect.py
from .now import now
from .Vec2 import Vec2

class MovableRect:

  def __init__(self, options={}):
    # Todo: Add validation
    self.width = options.get('width', 100)
    self.height = options.get('height', 100)
    self.startCenter = options.get('startCenter', Vec2(0, 0))
    self.startTime = options.get('startTime', now())
    self.endCenter = options.get('endCenter', self.startCenter.clone())
    self.endTime = options.get('endTime', self.startTime)

  def center(self, time=None):
    if time is None:
      time = now()

    if self.startTime >= self.endTime:
      return self.endCenter.clone()
    if time > self.endTime:
      return self.endCenter.clone()
    if time < self.startTime:
      return self.startCenter.clone()

    progress = (time - self.startTime) / (self.endTime - self.startTime)
    center = Vec2(
      self.startCenter.x + (self.endCenter.x - self.startCenter.x) * progress,
      self.startCenter.y + (self.endCenter.y - self.startCenter.y) * progress
    )

    return center

  def stop(self, time=None):
    if time is None:
      time = now()
    self.startCenter = self.center(time)
    self.startTime = time
    self.endCenter.copy(self.startCenter)
    self.endTime = time

  def top(self, time=None):
    if time is None:
      time = now()
    return self.center(time).y + self.height / 2

  def right(self, time=None):
    if time is None:
      time = now()
    return self.center(time).x + self.width / 2

  def bottom(self, time=None):
    if time is None:
      time = now()
    return self.center(time).y - self.height / 2

  def left(self, time=None):
    if time is None:
      time = now()
    return self.center(time).x - self.width / 2
  
  def json(self):
    return {
      'width': self.width,
      'height': self.height,
      'startCenter': self.startCenter.json(),
      'startTime': self.startTime,
      'endCenter': self.endCenter.json(),
      'endTime': self.endTime,
    }

File: front/app/src/game/utils/MovableRect.js
import Vec2 from './Vec2.js';

class MovableRect {
  constructor(options) {
    options = options || {};
    // todo: add validation
    this.width = options.width || 100;
    this.height = options.height || 100;
    this.startCenter = options.startCenter || new Vec2(0, 0);
    this.startTime = options.startTime || Date.now();
    this.endCenter = options.endCenter || this.startCenter.clone();
    this.endTime = options.endTime || this.startTime;
  }

  center(time) {
    time = time || Date.now();

    if (this.startTime >= this.endTime) return this.endCenter.clone();
    if (time > this.endTime) return this.endCenter.clone();
    if (time < this.startTime) return this.startCenter.clone();

    const progress = (time - this.startTime) / (this.endTime - this.startTime);
    const center = new Vec2(
      this.startCenter.x + (this.endCenter.x - this.startCenter.x) * progress,
      this.startCenter.y + (this.endCenter.y - this.startCenter.y) * progress
    );

    return center;
  }

  stop(time) {
    time = time || Date.now();
    this.startCenter = this.center(time);
    this.startTime = time;
    this.endCenter.copy(this.startCenter);
    this.endTime = time;
  }

  top(time) {
    time = time || Date.now();
    return this.center(time).y + this.height / 2;
  }

  right(time) {
    time = time || Date.now();
    return this.center(time).x + this.width / 2;
  }

  bottom(time) {
    time = time || Date.now();
    return this.center(time).y - this.height / 2;
  }

  left(time) {
    time = time || Date.now();
    return this.center(time).x - this.width / 2;
  }
}

export default MovableRect;
